backend/controller/evaljob_controller.go

```diff
@@ -0,0 +1,568 @@
+	"github.com/go-logr/logr"
+)
+
+const (
+	DriverPath                  = "/bin/driver"
+	DestDriverPath              = "/opt/app-root/src/bin/driver"
+	PodImageKey                 = "pod-image"
+	DriverImageKey              = "driver-image"
+	DriverServiceAccountKey     = "driver-serviceaccount"
+	DefaultPodImage             = "quay.io/yhwang/lm-eval-aas-flask:test"
+	DefaultDriverImage          = "quay.io/yhwang/lm-eval-aas-driver:test"
+	DefaultDriverServiceAccount = "driver"
+)
+
+// EvalJobReconciler reconciles a EvalJob object
+type EvalJobReconciler struct {
+	client.Client
+	Scheme    *runtime.Scheme
+	Recorder  record.EventRecorder
+	ConfigMap string
+	Namespace string
+	options   ServiceOptions
+}
+
+type ServiceOptions struct {
+	PodImage             string
+	DriverImage          string
+	DriverServiceAccount string
+}
+
+// +kubebuilder:rbac:groups=lm-eval-service.github.com,resources=evaljobs,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=lm-eval-service.github.com,resources=evaljobs/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=lm-eval-service.github.com,resources=evaljobs/finalizers,verbs=update
+// +kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch;create;delete
+// +kubebuilder:rbac:groups="",resources=configmaps,verbs=get;watch;list
+
+func (r *EvalJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	evalJob := &lmevalservicev1beta1.EvalJob{}
+	if err := r.Get(ctx, req.NamespacedName, evalJob); err != nil {
+		log.Error(err, "unable to fetch EvalJob. could be from an deletion request")
+		// we'll ignore not-found errors, since they can't be fixed by an immediate
+		// requeue (we'll need to wait for a new notification), and we can get them
+		// on deleted requests.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if !evalJob.ObjectMeta.DeletionTimestamp.IsZero() {
+		// Handle deletion here
+		return r.handleDeletion(ctx, evalJob, log)
+	}
+
+	// Treat this as NewJobState
+	if evalJob.Status.LastScheduleTime == nil {
+		evalJob.Status.State = lmevalservicev1beta1.NewJobState
+	}
+
+	// Handle the job based on its state
+	switch evalJob.Status.State {
+	case lmevalservicev1beta1.NewJobState:
+		// Handle newly created job
+		return r.handleNewCR(ctx, log, evalJob)
+	case lmevalservicev1beta1.ScheduledJobState:
+		// the job's pod has been created and the driver hasn't updated the state yet
+		// let's check the pod status and detect pod failure if there is
+		// TODO: need a timeout/retry mechanism here to transite to other states
+		return r.checkScheduledPod(ctx, log, evalJob)
+	case lmevalservicev1beta1.RunningJobState:
+		// TODO: need a timeout/retry mechanism here to transite to other states
+		return r.checkScheduledPod(ctx, log, evalJob)
+	case lmevalservicev1beta1.CompleteJobState:
+		return r.handleComplete(ctx, log, evalJob)
+	case lmevalservicev1beta1.CancelledJobState:
```

Comment:
Do we have cancellation separate from deletion? I could see that making sense so that the `status` and artifacts from a cancelled job. I think we may also want to consider some kind of retention policy for old jobs so that they don't pile up in the kube master.

Reply:
The `Cancelled` state is used to trigger the cancellation and is different from the `Complete` state in the current design. After a job is canceled, it will change to the `Complete` state and the reason will be `Cancelled`. Yes, we do need a retention policy but it could be either a separate mechanism from the controller or in the controller. We need to figure out a good way to consolidate reconcile-based and time-based logic.